storage/innobase/row/row0query.cc

```diff
@@ -0,0 +1,634 @@
+   dict_table_t *table, dtuple_t *search_tuple,
+   const upd_t *update, dtuple_t *insert_tuple) noexcept
+{
+retry_again:
+  dberr_t err= select_for_update(table, search_tuple);
+  if (err == DB_SUCCESS)
+  {
+    err= update_record(table, update);
+    if (err == DB_SUCCESS) err= commit_update();
+    else rollback_update();
+    return err;
+  }
+  else if (err == DB_RECORD_NOT_FOUND)
+  {
+    if (m_mtr_active)
+    {
+      m_mtr.commit();
+      m_mtr_active= false;
+    }
+    err= insert_record(table, insert_tuple);
+    return err;
+  }
+  else if (err == DB_LOCK_WAIT)
+    goto retry_again;
+  return err;
+}
+
+dberr_t QueryExecutor::read(dict_table_t *table, dtuple_t *tuple,
+                            page_cur_mode_t mode,
+                            RecordCallback& callback) noexcept
+{
+  ut_ad(table);
+  dict_index_t *index= dict_table_get_first_index(table);
+  if (!index) return DB_ERROR;
+
+  m_mtr.start();
+  if (m_trx && !m_trx->read_view.is_open())
+  {
+    trx_start_if_not_started(m_trx, false);
+    m_trx->read_view.open(m_trx);
+  }
+  m_pcur.btr_cur.page_cur.index= index;
+  dberr_t err= DB_SUCCESS;
+  if (tuple) err= btr_pcur_open(tuple, mode, BTR_SEARCH_LEAF, &m_pcur, &m_mtr);
+  else
+  {
+    err= m_pcur.open_leaf(true, index, BTR_SEARCH_LEAF, &m_mtr);
+    if (err == DB_SUCCESS) btr_pcur_move_to_next(&m_pcur, &m_mtr);
+  }
+  if (err != DB_SUCCESS)
+  {
+    m_mtr.commit();
+    return err;
+  }
+  ulint match_count= 0;
+  while (btr_pcur_is_on_user_rec(&m_pcur))
+  {
+    const rec_t *rec= btr_pcur_get_rec(&m_pcur);
+    rec_offs* offsets= rec_get_offsets(rec, index, nullptr,
+                                       index->n_core_fields,
+                                       ULINT_UNDEFINED, &m_heap);
+    RecordCompareAction action= callback.compare_record(
+      tuple, rec, index, offsets);
+    if (action == RecordCompareAction::PROCESS)
+    {
+      bool continue_processing= process_record_with_mvcc(table, index, rec,
+                                                         offsets, callback,
+                                                         &m_mtr, match_count);
+      if (!continue_processing) break;
+    }
+    else if (action == RecordCompareAction::STOP)
+      break;
+    if (!btr_pcur_move_to_next(&m_pcur, &m_mtr)) break;
+  }
+  m_mtr.commit();
+  return (match_count > 0 || !tuple) ? DB_SUCCESS : DB_RECORD_NOT_FOUND;
+}
+
+dberr_t QueryExecutor::read_by_index(dict_table_t *table,
+                                     dict_index_t *sec_index,
+                                     dtuple_t *search_tuple,
+                                     page_cur_mode_t mode,
+                                     RecordCallback& callback) noexcept
+{
+  ut_ad(table);
+  ut_ad(sec_index);
+  ut_ad(sec_index->table == table);
+  ut_ad(!dict_index_is_clust(sec_index));
+
+  dict_index_t *clust_index= dict_table_get_first_index(table);
+  if (!clust_index) return DB_ERROR;
```

Comment:
The function should be declared with `__attribute__((nonnull))` and the related assertions removed. Why is `search_tuple` not a pointer to `const`? There is a space before each unary `*` but after the unary `&`. That is inconsistent.

Do we really need the `if` condition here? There should be no code path that would add a table to `dict_sys` such that it does not have any indexes.